lib/async_hooks.js

```diff
@@ -204,11 +206,145 @@ class AsyncResource {
   }
 }
 
+let activeAsyncLocalStorageCount = 0;
+const kLocalStorageList = Symbol('kLocalStorageList');
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const currentResource = executionAsyncResource();
+    // Value of triggerResource is always a non null object
+    const localStorageList = currentResource[kLocalStorageList];
+    if (!localStorageList) {
+      return;
+    }
+    for (let i = 0; i < localStorageList.length; ++i) {
+      localStorageList[i]._propagate(resource, currentResource);
+    }
+  }
+});
+
+class AsyncLocalStorage {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+    this.enabled = false;
+  }
+
+  enable() {
+    if (!this.enabled) {
+      this.enabled = true;
+      ++activeAsyncLocalStorageCount;
+      asyncContextHook.enable();
+      const resource = executionAsyncResource();
+      let localStorageList = resource[kLocalStorageList];
+      if (!localStorageList) {
+        localStorageList = resource[kLocalStorageList] = [];
+      }
+      localStorageList.push(this);
+    }
+  }
+
+  disable() {
+    if (this.enabled) {
+      this.enabled = false;
+      --activeAsyncLocalStorageCount;
+      const localStorageList = executionAsyncResource()[kLocalStorageList];
+      if (localStorageList) {
+        localStorageList.splice(localStorageList.indexOf(this), 1);
+      }
+      if (activeAsyncLocalStorageCount === 0) {
+        asyncContextHook.disable();
+      }
+    }
+  }
+
+  /**
+   * Propagate the context from a parent resource to a child one.
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (this.enabled) {
+      resource[this.kResourceStore] = store;
+      let list = resource[kLocalStorageList];
+      if (!list) {
+        list = resource[kLocalStorageList] = [];
+      }
+      list.push(this);
```

Comment:
This approach may lead to mem leaks in case of reused async resource objects, as there is no check for existence of `AsyncLocalStorage` instance in the array. Thus each time when the async resource object is reused (which may happen at least with user land modules and some of core modules) the instance will be pushed to the array, even if it already was there. I may be wrong here, but I don't see anything that prevents such scenario in the code.

Reply:
That's true. The change suggested in https://github.com/nodejs/node/pull/26540#discussion_r379922059 should fix this right?

Reply:
It should, but that depends on the exact fix. So, it's better to keep that in mind.

Reply:
Current solution will work as there is an `enabled` check before pushing to the list

Reply:
Unfortunately, `enabled` check doesn't protect from duplicates in the list, even when async resources aren't reused. Here is a minimal test for an illustration of the problem:
```js
'use strict';
require('../common');
const assert = require('assert');
const { AsyncLocalStorage, executionAsyncResource } = require('async_hooks');

const asyncLocalStorage = new AsyncLocalStorage();

// branch one
setTimeout(() => {
  asyncLocalStorage.enable();

  setTimeout(() => {
    asyncLocalStorage.runSyncAndReturn(() => {
      // a hacky way to obtain list of storages within this branch
      const symbols = Object.getOwnPropertySymbols(executionAsyncResource());
      const listSymbol = symbols[symbols.length - 1];
      const localStorageList = executionAsyncResource()[listSymbol];

      assert.strictEqual(localStorageList[0], asyncLocalStorage);
      // this one fails because there are two elements in the array
      assert.strictEqual(localStorageList.length, 1);
    });
  }, 200);
}, 0);

// branch two
setTimeout(() => {
  asyncLocalStorage.disable();
}, 100);
```

As lists are stored in symbol properties local to each async call chain, the `.disable()` call here doesn't remove the storage object from the list in branch one.

This wouldn't happen with the old (global) list of storage instances. So, maybe it's worth considering a move back to that approach?